sdk/include/opentelemetry/sdk/trace/multi_span_processor.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <mutex>
+#include <vector>
+
+#include "opentelemetry/sdk/trace/multi_recordable.h"
+#include "opentelemetry/sdk/trace/processor.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+
+/** Instantiation options. */
+struct MultiSpanProcessorOptions
+{};
+
+/**
+ * Span processor allow hooks for span start and end method invocations.
+ *
+ * Built-in span processors are responsible for batching and conversion of
+ * spans to exportable representation and passing batches to exporters.
+ */
+class MultiSpanProcessor : public SpanProcessor
+{
+public:
+  MultiSpanProcessor(std::vector<std::unique_ptr<SpanProcessor>> &&processors)
+  {
+    for (auto &processor : processors)
+    {
+      if (processor)
+      {
+        processors_.push_back(std::move(processor));
+      }
+    }
+  }
+
+  void AddProcessor(std::unique_ptr<SpanProcessor> &&processor)
```

Comment:
Good comment. My earlier code did have mutex guard against all vector operations in this class, seems somehow missed out check-in it The doubly linked list looks like a good option too, would be good to do some benchmarking to select the optimal one of two.

Reply:
I can share the thinking from the .NET implementation:

STL vector + mutex would give the ultimate thread-safety in multi-reader-multi-writer situation, the potential downside is that it will introduce contention to the readers (in this case, the readers are the threads that push spans to the processor chain), which might something to avoid.

Doubly linked list only gives single-writer-multi-reader thread-safety, however it is well aligned with the general .NET API guideline (that if a class method doesn't specify thread safety, by default it is not supposed to be called concurrently).

Reply:
Ok. now I realize why I removed the mutex guard - as `MultiSpanProcessor` is an atomic instance in `TracerContext` :

https://github.com/open-telemetry/opentelemetry-cpp/blob/f10acac2fe708635dfb0b48760bc598270b832b9/sdk/include/opentelemetry/sdk/trace/tracer_context.h#L86

And so multiple threads shouldn't be able to access it simultaneously. But will wait for comments on others :)

Reply:
Atomic just means that you can compare-and-swap the pointer.  Users can still access it simultaneously (after reading the pointer in memory).

*if* you have `AddProcessor` return a new MultiProcessor instance and compare-and-swap the Atomic Ptr then I think you'd be ok, but as is, you still have a (possible) threading issue if more than thread calls AddProcessor.  It's unlikely but still possible.

Reply:
OK, I think I got your point, in that case, we should have Atomic Ptr for Span Processor, and create new instance of MultiProcessor everytime AddProcessor() is called. It is already documented that AddProcessor() is not thread safe, and shouldn't be called simultaneously.

Reply:
Ah, I missed that AddProcessor was marked not thread-safe as well.   This means my point is not important.

If the changes to make AddProcessor thread-safe aren't too painful, I think it's worth it.  I think it's a low-frequency method call, so it's ok to be "expensive" to AddProcessor as long as the "read" hot path is fast (which you already have AFAICT).   Removing my requested changes.  I'd prefer to see AddProcessor thread-safe but this is fine as-is.